railgun/apis/configuration/v1beta1/ingress_rules.go

```diff
@@ -0,0 +1,46 @@
+package v1beta1
+
+// UDPIngressRule represents a rule to apply against incoming requests
+// wherein no Host matching is available for request routing, only the port
+// is used to match requests.
+type UDPIngressRule struct {
+	// Port indicates the port for the Kong proxy to accept incoming traffic
+	// on, which will then be routed to the service Backend.
+	Port int `json:"port"`
+
+	// Backend defines the Kubernetes service which accepts traffic from the
+	// listening Port defined above.
+	Backend IngressBackend `json:"backend"`
+}
+
+// IngressRule represents a rule to apply against incoming requests.
+// Matching is performed based on an (optional) SNI and port.
+type IngressRule struct {
+	// Host is the fully qualified domain name of a network host, as defined
+	// by RFC 3986.
+	// If a Host is specified, the protocol must be TLS over TCP.
+	// A plain-text TCP request cannot be routed based on Host. It can only
+	// be routed based on Port.
+	// +optional
+	Host string `json:"host,omitempty"`
+
+	// Port is the port on which to accept TCP or TLS over TCP sessions and
+	// route. It is a required field. If a Host is not specified, the requested
+	// are routed based only on Port.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Port int `json:"port,omitempty"`
+
+	// Backend defines the referenced service endpoint to which the traffic
+	// will be forwarded to.
+	Backend IngressBackend `json:"backend"`
+}
+
+// IngressBackend describes all endpoints for a given service and port.
+type IngressBackend struct {
```

Comment:
Consider KIC 2.0 one-upping 1.x by using `IngressBackend` from `k8s.io/api@v0.21.1/networking/v1beta1` instead for the following benefits:

- consistency with `Ingress` API
- possibility (current or future) to reference service ports by name
- possibility (current or future) to reference objects other than services

Since the API version here is already beta, we should strive for the best API we can think of from the get-go because any breaking changes will be painful and costly in the longer run.

This change will require an enhancement of `fromIngressV1beta1` to support the case of ServicePort being a string. We have prior art: 

https://github.com/Kong/kubernetes-ingress-controller/blob/d764097011290204e01038228cd7c1ea204ad3b0/pkg/parser/translate.go#L112

and

https://github.com/Kong/kubernetes-ingress-controller/blob/d764097011290204e01038228cd7c1ea204ad3b0/pkg/parser/parser.go#L162

No need to do this as part of this PR but let's consider this a potential improvement to be shipped with KIC 2.0.

Reply:
I have created the following issue as a follow up for this:

https://github.com/Kong/kubernetes-ingress-controller/issues/1428

Reply:
Sounds good. Reworded the issue because I believe it didn't capture the functional part of the motivation.